log/file.go

```diff
@@ -0,0 +1,157 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+// defaultLogfilePath is a default logfile path.
+const defaultLogfilePath = "/tmp/k6run.log"
+
+// fileHook is a hook to handle writing to local files.
+type fileHook struct {
+	fallbackLogger logrus.FieldLogger
+	ch             chan []byte
+	path           string
+	w              io.WriteCloser
+	levels         []logrus.Level
+}
+
+// FileHookFromConfigLine returns new fileHook hook.
+func FileHookFromConfigLine(
+	ctx context.Context, fallbackLogger logrus.FieldLogger, line string,
+) (logrus.Hook, error) {
+	hook := &fileHook{
+		fallbackLogger: fallbackLogger,
+		ch:             make(chan []byte),
+		path:           defaultLogfilePath,
+		levels:         logrus.AllLevels,
+	}
+
+	if fallbackLogger != nil {
+		hook.fallbackLogger = fallbackLogger
+	}
+
+	if line != "file" {
+		parts := strings.SplitN(line, "=", 2)
+		if parts[0] != "file" {
+			return nil, fmt.Errorf("logfile configuration should be in the form `file=path-to-local-file` but is `%s`", line)
+		}
+
+		if err := hook.parseArgs(line); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := hook.createOpenFile(); err != nil {
+		return nil, err
+	}
+
+	go hook.loop(ctx)
+
+	return hook, nil
+}
+
+func (h *fileHook) parseArgs(line string) error {
+	tokens, err := tokenize(line)
+	if err != nil {
+		return fmt.Errorf("error while parsing logfile configuration %w", err)
+	}
+
+	for _, token := range tokens {
+		switch token.key {
+		case "file":
+			if token.value == "" {
+				return fmt.Errorf("filepath must not be empty")
+			}
+			h.path = token.value
+		case "level":
+			h.levels, err = parseLevels(token.value)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown logfile config key %s", token.key)
+		}
+	}
+
+	return nil
+}
+
+// createOpenFile opens logfile by provided path,
+// it will create a file if it doesn't exist.
+func (h *fileHook) createOpenFile() error {
```

Comment:
>I find the name of this function quite unconventional

IMHO, it would make sense in the case it would be an exported method so a public API for this struct. But it's something internal for wrapping the `OR` logic. If we delegate the input again then better to have directly the entire logic in the caller, otherwise we have just one more layer of logic where we could generate a failure.

Maybe we can improve the name? Just a suggestion, I don't have one better at the moment.

```suggestion
func (h *fileHook) openOrCreateFile() error {
```

Reply:
I've given this a bit more thought and, given the choice, I do think I'd prefer to get rid of this helper function all in all.

If we were to keep the function, I find that the naming is not the issue. Whether you open, or create the file ends up in the same underlying standard library call, and system call. It's always technically opening a file, with, in one scenario the instruction to not fail if it doesn't exist, but have the OS create it instead. But I'm really, REALLY, nitpicking on this, and I really don't want it to be a blocker! (from my point of view, it's not)